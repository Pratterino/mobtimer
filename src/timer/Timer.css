.timer {
    border: 2px solid brown;
    display: inline-flex;
    align-items: center;
    justify-content: center;
}

.timer__time {
    position: absolute;
    font-size: 40px;
}
