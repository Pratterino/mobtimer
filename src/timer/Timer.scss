$transitionTime: 0.2s;
$activeColor: #66cdaa;
$inactiveColor: #d7686c;
$pausedColor: #d7d283;

.Timer {
    margin: 40px 0;

    .timer {
        box-sizing: content-box;
        display: inline-flex;
        align-items: center;
        justify-content: center;

        .timer__time {
            text-shadow: 2px 2px 2px #000;
            user-select: none;
            color: #f9f9f9;
            position: absolute;
            font-size: 50px;
        }

        .paused,
        .inactive,
        .active {
            transition: stroke $transitionTime ease-in-out;
        }

        .timer__time--numbers {
            display: block;
        }

        .timer__time--icons {
            display: none;
        }

        &:hover {
            .timer__time--numbers {
                display: none;
            }

            .timer__time--icons {
                display: block;
                -webkit-filter: drop-shadow(2px 2px 2px #000);
                filter: drop-shadow(2px 2px 2px #000);
            }
        }
    }
}
