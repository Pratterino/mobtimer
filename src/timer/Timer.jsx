import React, {Component} from 'react';
import {connect} from "react-redux";
import {bindActionCreators} from "redux";
import {playPauseTimer, startTimer, stopTimer} from "./timerActions";
import {getParsedTimeRemaining, lightenDarkenColor} from "./../helper/TimerHelper";
import {FontAwesomeIcon} from '@fortawesome/react-fontawesome'
import {faPause, faPlay} from '@fortawesome/free-solid-svg-icons'
import "./Timer.scss";
import "./TimerCircle.scss";

class Timer extends Component {
    getPercentageLeftOfTime = () => {
        const {currentTime, sessionLength} = this.props.timer;
        return (currentTime / sessionLength) * 100;
    };

    renderCircularProgressbar = () => {
        const circleGradient = this.props.timer.active ? "active" : "inactive";
        const backgroundShadowColor = "rgba(25, 25, 25, 0.5)";
        const backgroundColor = getComputedStyle(document.body).getPropertyValue('--background');
        const activeColor = getComputedStyle(document.body).getPropertyValue('--active-timer-color');
        const stopColor = getComputedStyle(document.body).getPropertyValue('--stopped-timer-color');
        const luminance = 0.5;

        return (
            <svg className="circle-chart"
                 viewBox="0 0 33.83098862 33.83098862"
                 width="180"
                 height="180"
                 xmlns="http://www.w3.org/2000/svg">
                <defs>
                    <linearGradient id="background-shadow" x1="0%" y1="0%" x2="0%" y2="100%">
                        <stop offset="0%" stopColor={backgroundShadowColor}/>
                    </linearGradient>
                    <linearGradient id="background" x1="0%" y1="0%" x2="0%" y2="100%">
                        <stop offset="0%" stopColor={backgroundColor}/>
                        <stop offset="100%" stopColor={lightenDarkenColor(backgroundColor, luminance)}/>
                    </linearGradient>
                    <linearGradient id="inactive" x1="0%" y1="0%" x2="0%" y2="100%">
                        <stop offset="0%" stopColor={stopColor}/>
                        <stop offset="100%" stopColor={lightenDarkenColor(stopColor, luminance)}/>
                    </linearGradient>
                    <linearGradient id="active" x1="0%" y1="0%" x2="0%" y2="100%">
                        <stop offset="0%" stopColor={activeColor}/>
                        <stop offset="100%" stopColor={lightenDarkenColor(activeColor, luminance)}/>
                    </linearGradient>
                </defs>
                <circle className="circle-chart__background-shadow"
                        stroke="url(#background-shadow)"
                        strokeWidth="4"
                        fill="none"
                        cx="16.91549431"
                        cy="16.91549431"
                        r="15.91549431"
                />
                <circle className="circle-chart__background"
                        stroke="url(#background)"
                        strokeWidth="3"
                        fill="none"
                        cx="16.91549431"
                        cy="16.91549431"
                        r="15.91549431"
                />
                <circle className={`circle-chart__circle`}
                        stroke={`url(#${circleGradient})`}
                        strokeWidth="3"
                        strokeDasharray={`${this.getPercentageLeftOfTime() || 100},100`}
                        strokeLinecap="round"
                        fill="none"
                        cx="16.91549431"
                        cy="16.91549431"
                        r="15.91549431"
                />
            </svg>
        );
    };

    renderTimeRemaining = (seconds) => {
        const time = getParsedTimeRemaining(seconds);

        return (
            <div className="timer__time">
                <div className="timer__time--numbers">
                    {time}
                </div>
                <div className="timer__time--icons">
                    <FontAwesomeIcon icon={this.props.timer.active ? faPause : faPlay}/>
                </div>
            </div>
        );
    };

    render() {
        const {settings, timer} = this.props;
        return (
            <section className="Timer">
                <div
                    className="timer pointer"
                    onClick={this.props.playPauseTimer.bind(null, this.props.timer)}
                >
                    {this.renderCircularProgressbar()}
                    {this.renderTimeRemaining(timer.currentTime)}
                </div>
            </section>
        );
    }
}

const mapStateToProps = state => ({
    timer: state.timer,
});
const mapDispatchToProps = dispatch => bindActionCreators({
    stopTimer,
    playPauseTimer,
    startTimer,
}, dispatch);

export const UnwrappedTimer = Timer;
export default connect(mapStateToProps, mapDispatchToProps)(Timer);
