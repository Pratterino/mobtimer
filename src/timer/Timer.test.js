import React from 'react';
import { unwrapped as UnwrappedTimer } from './Timer';
import { shallow } from 'enzyme';

describe('Timer', () => {
    let wrapper;

    const renderComponent = props => {
        wrapper = shallow(
            <UnwrappedTimer
                timer={{
                    active: false,
                    currentTime: 100,
                    sessionLength: 100,
                }}
                stopTimer={jest.fn()}
                playPauseTimer={jest.fn()}
                startTimer={jest.fn()}
                {...props}
            />,
        );
    };

    beforeEach(() => {
        renderComponent();
    });

    describe('renderTimeRemaining', () => {});

    describe('render', () => {
        it('should match snapshot', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });
});
