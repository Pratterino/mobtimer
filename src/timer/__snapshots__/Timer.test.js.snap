// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer render should match snapshot 1`] = `
<section
  className="Timer"
>
  <div
    className="pointer timer"
    onClick={[Function]}
  >
    <TimerCircle
      timer={
        Object {
          "active": false,
          "currentTime": 100,
          "sessionLength": 100,
        }
      }
    />
  </div>
</section>
`;
